src/transforms/window.js

```diff
@@ -120,66 +136,144 @@ function reduceSum(k, s) {
         const b = S[I[i + k - 1]];
         if (b === null || isNaN(b)) ++nans;
         else sum += +b;
-        T[I[i + s]] = nans === 0 || skipNaN ? sum : NaN;
+        T[I[i + s]] = nans === 0 ? sum : NaN;
         if (a === null || isNaN(a)) --nans;
         else sum -= +a;
       }
     }
-  };
+  }) : ({
+    map(I, S, T) {
+      let sum = 0;
+      const n = I.length;
+      for (let i = 0, j = Math.min(n, k - s - 1); i < j; ++i) {
+        sum += +S[I[i]] || 0;
+      }
+      for (let i = -s, j = n - s; i < j; ++i) {
+        sum += +S[I[i + k - 1]] || 0;
+        T[I[i + s]] = sum;
+        sum -= +S[I[i]] || 0;
+      }
+    }
+  });
 }
 
-function reduceMean(k, s,) {
-  const sum = reduceSum(k, s);
-  return {
-    map(I, S, T, skipNaN) {
-      sum.map(I, S, T, skipNaN);
-      for (let i = 0, n = I.length - k + 1; i < n; ++i) {
-        T[I[i + s]] /= k;
+function reduceMean(k, s, strict) {
+  if (strict) {
+    const sum = reduceSum(k, s, strict);
+    return {
+      map(I, S, T) {
+        sum.map(I, S, T);
+        for (let i = 0, n = I.length - k + 1; i < n; ++i) {
+          T[I[i + s]] /= k;
+        }
       }
-    }
-  };
+    };
+  } else {
+    return {
+      map(I, S, T) {
+        let sum = 0;
+        let count = 0;
+        const n = I.length;
+        for (let i = 0, j = Math.min(n, k - s - 1); i < j; ++i) {
+          let v = S[I[i]];
+          if (v !== null && !isNaN(v = +v)) sum += v, ++count;
+        }
+        for (let i = -s, j = n - s; i < j; ++i) {
+          let a = S[I[i + k - 1]];
+          let b = S[I[i]];
+          if (a !== null && !isNaN(a = +a)) sum += a, ++count;
+          T[I[i + s]] = sum / count;
+          if (b !== null && !isNaN(b = +b)) sum -= b, --count;
+        }
+      }
+    };
+  }
+}
+
+function valid(x) {
+  return x !== null && !isNaN(x);
```

Comment:
This check is appropriate if we are going to coerce to a number (for the _ratio_ and _difference_ reducers), but if we’re not (for the _first_ and _last_ reducers), we want to use `defined` and `Number.isNaN` instead of `isNaN`. I will fix.